### DISCUSSION TITLE
tegrademo-mender setting rootfs size without changing sd/emmc size

### DISCUSSION THREAD

**Discussion by goetzdd** [OE4T/tegra-demo-distro#50] [Q&A] [answered]
Ok, I'm very new to yocto/mender but have worked with the Jetson-TX1/TX2 previously.

I modified the flash layout within the nvidia SDK environment and changed the rootfs size to allow for adding some partitions to the device, but I'm struggling to to do the same using the tegrademo-mender distro.

Basically I need to reduce the size of the rootfs (APP/APP_b) partitions to make space for the additional partitions I'd like to add. I've see the discussion/issue that changes the SD/emmc size, but I can't seem to find the special sauce to just reduce the rootfs sizes using the same SD/emmc size.

Also to start with I'm just leaving the space unpartitioned, but i have existing sparse .img I assume I could use, but the very basics of yocto are messing with my head at this point. I've created a meta-xxx layer for my specific stuff, but files and recipes and bbappends, oh my...

And to make matters more confused (more correctly, make me more confused) the aim is to build tx2 and xavier-nx machines to start, but preferably later be able to add meta-xilinx and bring my modifications accomplished for nvidia right along to building for zynqmp. Looking at the structure of the tegra-demo-distro it kind of looks like I should start from scratch and build the yocto project folder more in line with the normal distro and manually build up the layers used in tegra-demo-distro so when I then add meta-xilinx I can proceed a bit more along what looks like normal yocto flow. Having become acquainted with Petalinux a bit I'd rather maintain everything in one place rather than all over the map.

On edit: Sorry I think I should have put this in tegra-demo-distro, not meta-tegra

**Comment by madisongh** [comment]
IIRC, the variable you're looking for `MENDER_IMAGE_ROOTFS_SIZE` (which I think is in KiB).  But I have the same issue trying to figure out how to deal with the conflict between the BSP tools wanting to control the flash layout and partition sizes vs. Mender wanting to do so, and usually have to experiment around multiple times to get things to work out right.

**Reply by goetzdd** [reply]
I ended up using TEGRA_MENDER_RESERVED_SPACE_MB, but make it too large and you get an error (filesystem too large, but it looks like its a math error and somewhere generates a negative number to compare to IMAGE_ROOTFS_MAXSIZE) . I grepped for MENDER_IMAGE_ROOTFS_SIZE and only find MENDER_IMAGE_ROOTFS_SIZE_DEFAULT, possibly this is what you mean?

**Comment by dwalkes** [comment, marked as answer]
> Looking at the structure of the tegra-demo-distro it kind of looks like I should start from scratch and build the yocto project folder more in line with the normal distro and manually build up the layers used in tegra-demo-distro so when I then add meta-xilinx I can proceed a bit more along what looks like normal yocto flow.

I'm not sure what this means but I'm interested in any way we could make tegra-demo-distro work better in this scenario, or what specific things aren't going to work well for adding meta-xilinx (which I haven't tried to use myself).

> I've see the discussion/issue that changes the SD/emmc size, but I can't seem to find the special sauce to just reduce the rootfs sizes using the same SD/emmc size.

Agree this is confusing, I've upated [this page](https://github.com/OE4T/tegra-demo-distro/wiki/Customization-of-Storage-Size-and-root-filesystem-size-on-Mender-Builds) to discuss how we solve this in more detail, there may be better/alternate ways.  It would be great to get some different approaches documented.

> On edit: Sorry I think I should have put this in tegra-demo-distro, not meta-tegra

No worries, Matt may even be able to move these?  I can't tell if Github has this feature for discussions or only for issues?

**Reply by madisongh** [reply]
> No worries, Matt may even be able to move these? I can't tell if Github has this feature for discussions or only for issues?

Yep, looks like it worked.

**Reply by goetzdd** [reply]
> > Looking at the structure of the tegra-demo-distro it kind of looks like I should start from scratch and build the yocto project folder more in line with the normal distro and manually build up the layers used in tegra-demo-distro so when I then add meta-xilinx I can proceed a bit more along what looks like normal yocto flow.
> 
> I'm not sure what this means but I'm interested in any way we could make tegra-demo-distro work better in this scenario, or what specific things aren't going to work well for adding meta-xilinx (which I haven't tried to use myself).

Compared to the examples all the meta-xxx layers are in their own <workdir>/layers with links to <workdir>/repos, maybe I'm making things too difficult though. I imagine I can add meta-xilinx in repos and create a link, or directly in layers, or any number of things I haven't discovered yet.

My main point was building my own distro setup for our usage, but I'm not quite that smart about yocto yet. In some ways working with this seems somewhat backwards, you experiment until it does what you want and you understand it better and then start over to make it lean, clean and organized.

> 
> > I've see the discussion/issue that changes the SD/emmc size, but I can't seem to find the special sauce to just reduce the rootfs sizes using the same SD/emmc size.
> 
> Agree this is confusing, I've upated [this page](https://github.com/OE4T/tegra-demo-distro/wiki/Customization-of-Storage-Size-and-root-filesystem-size-on-Mender-Builds) to discuss how we solve this in more detail, there may be better/alternate ways. It would be great to get some different approaches documented.
> 
> > On edit: Sorry I think I should have put this in tegra-demo-distro, not meta-tegra
> 
> No worries, Matt may even be able to move these? I can't tell if Github has this feature for discussions or only for issues?

**Reply by dwalkes** [reply]
> I imagine I can add meta-xilinx in repos and create a link, or directly in layers, or any number of things I haven't discovered yet.

Yea I think the world's your oyster here.  I've tried a few different setups but I like the format with the "repos" subdir for repositories and then links from layers.  This lets you easily see all the different modules you are cloning and inventory their upstream hashes in one spot, and also allows you to filter on grep searches from the links directory to remove things you aren't actually using.  This came from @madisongh like most of the content here, and I'm not sure where or if it originated somewhere outside his head.

> but I'm not quite that smart about yocto yet

Me too!

> you experiment until it does what you want and you understand it better and then start over to make it lean, clean and organized.

There's some truth to that for sure.  I've definitely found learning by doing to be a good approach, and there's plenty to learn from in these repos and thankfully a community of people smarter than me to learn from.
